Fix and improve 'show upf session'

ivan4th opened the pull request:
Fix filtering by hex UP SEIDs (`show upf session up seid 0x...`)
Add `show upf session 0x... flows` to show the flows related to the
session.
Limit `show upf session` to displaying 100 sessions by default.
This can be overridden via `show upf session limit N`
(`limit 0` stands for unlimited)

hwinkel commented:
Do we have by IMSI as well, as we signal the imsi now down to UPG  from SMC As part of PFCP IEs

RoadRunnr commented:
> Do we have by IMSI as well, as we signal the imsi now down to UPG from SMC As part of PFCP IEs

This would require additional hashes to index the session by IMSI/IMEI/MSISDN/what-not. The overhead and potential for problems is IMHO not worth it.

RoadRunnr commented on upf/upf_cli.c:
```diff
@@ -1197,7 +1221,7 @@ VLIB_CLI_COMMAND (upf_show_session_command, static) =
   .path = "show upf session",
   .short_help =
-  "show upf session",
+  "show upf session [up seid 0x...] [0x... flows] [limit N]",
```
do not mix prefix and postfix parameters. VPP cli command always use the prefix form, stick with that please.

The VPP CLI is not operation tool! The usability of the command is not really important. But, you create a potential of messing up the command parsing if you mix postfix and prefix arguments like that.

RoadRunnr replied:
Also, limiting the output is pretty useless. Either you know that it is safe to list all or there are too many session that listing even a part of them is useless anyhow.

I would suggest to change the command to either display a single session or all. Use a special flag to force all, e.g.:
```
"show upf session [ up seid N [flows N] | all]"
```

korroot replied:
If we have like 1k sessions and want to limit the output, so running like `show upf session limit 100` will always show us first 100 sessions, and the rest 900 will not be accessible - either there should be paging (like `limit 100-200`) or just remove the limit... 

But maybe if first entries can say us something meaningful than it can stay, what we can gain by seeing first sessions?

korroot replied:
When I reread the commit description the help should be:
```
show upf session [up seid 0x...][up seid 0x... flows][limit N]
```
Right?

ivan4th commented:
Well, risk-wise a dumber approach of traversing the whole session pool looking for a particular IMSI might not be that bad, if it is not done repeatedly in a tight loop, which typically doesn't happen during CLI debug sessions.

From the experience, `show upf flows` doesn't cause disruptions most of the time despite going over the whole flowtable, and the session pool usually contains fewer entries. The risk of the disruption from `show upf session` stems from generating the output, not from examining the data structures.

sergeymatov reviewed (approved):
lgtm

ivan4th commented:
Updated the PR. The syntax now is
```
show upf session [up seid 0x... [flows]] [limit N]
```

Indeed, I think showing just 100 sessions are still useful b/c of 2 possibilities
1) pure debug/test settings where the actual number of sessions is less than 100
2) to get some clues from a number of sessions in the system

And having the limit in place by default is also critical b/c `show upf session` is known to have caused actual UPG crashes in prod due to the amount of output.

I've also added `limit N` option to `show upf flows`. There's no limit by default b/c as far as I know, UPG never ever has crashed in prod when executing this command. But the limit may be useful for synthetic situations where unusually high number of flows is created using TPLoad or TRex, for example.

RoadRunnr commented on upf/upf_cli.c:
```diff
@@ -1108,34 +1121,33 @@ upf_show_session_command_fn (vlib_main_t * vm,
 	    debug = 1;
-#if FLOWTABLE_TODO
-	  else if (unformat (line_input, "%lu flows", &up_seid))
+	  else if (unformat (line_input, "0x%llx flows", &up_seid))
```
what is this used for? all cases for parsing up_seid should already be covered.

RoadRunnr commented on upf/upf_cli.c:
```diff
@@ -1147,28 +1159,45 @@ upf_show_session_command_fn (vlib_main_t * vm,
+      u32 cpu_index;
+      flowtable_main_t *fm = &flowtable_main;
+      vlib_thread_main_t *tm = vlib_get_thread_main ();
+      flows_out_arg_t arg = {.vm = vm,.seid = up_seid,.filtered =
+	  true,.limit = limit
+      };
```
Strange that this passes through the format checks, I would expected that there should be a space after a comma.

ivan4th replied:
Hmm, it looks like it was actually `indent` which formatted it this way, and I overlooked it. Split it into separate lines

RoadRunnr commented on upf/upf_cli.c:
```diff
@@ -1108,34 +1121,33 @@ upf_show_session_command_fn (vlib_main_t * vm,
+	  else if (unformat (line_input, "flows", &up_seid))
 	    has_flows = 1;
-#endif
+	  else if (unformat (line_input, "limit %u", &limit))
+	    ;
```
should we enforce a hard max for limit? Currently setting limit to 0 would effectively mean `unlimited`

ivan4th replied:
Added the hard limit of 10K sessions
